server/src/main/java/org/elasticsearch/action/bulk/IncrementalBulkService.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.bulk;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.support.ActiveShardCount;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.bytes.CompositeBytesReference;
+import org.elasticsearch.common.bytes.ReleasableBytesReference;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.rest.BaseRestHandler;
+import org.elasticsearch.rest.RestChannel;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.search.fetch.subphase.FetchSourceContext;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+public class IncrementalBulkService {
+
+    private final boolean allowExplicitIndex;
+    private final Client client;
+
+    public IncrementalBulkService(Settings settings, Client client) {
+        this.allowExplicitIndex = BaseRestHandler.MULTI_ALLOW_EXPLICIT_INDEX.get(settings);
+        this.client = client;
+    }
+
+    public BaseRestHandler.RequestBodyChunkConsumer newBulkRequest(RestRequest request, ActionListener<BulkResponse> listener) {
```

Comment:
I wonder if we should decouple this from the Rest layer? That would also allow you to work on this more independently - and seems it would allow easier IT too?

I think we could have `newBulkRequest(args)` returning a `IncrementalBulkHandler`, that has `addItem(DocWriteRequest)`. Handling parsing and chunks would thus reside outside this.

I'd at least like to split into those two concerns being handled separately (incrementally writing data when necessary and the rest chunk parsing, pausing etc).

Reply:
Not sure if it's what Henning commented on. I think any bulk processor would like to split chunk by separator. Interface that would indicate end of bulk item might be handy.

```java
    interface BulkChunkConsumer {
        void onNext(ReleasableBytesReference chunk, boolean itemEnd, boolean streamEnd);
    }
```
Where implementation will wrap raw network chunks and slice them by bulk separator, no aggregation. Roughly this. This slicer will keep single netty buffer at most in memory and slice it (may be many times), when current is exhausted request next from the upstream.

```java
    class BulkChunkSlicer {
        HttpBody.Stream rawStream;
        String separator;
        ReleasableBytesReference current;
        boolean streamEnd;
        BulkChunkConsumer consumer;

        BulkChunkSlicer(HttpBody.Stream rawStream) {
            rawStream.setHandler((chunk, isLast) -> {
                current = chunk;
                streamEnd = isLast;
                if (canSliceCurrent()) {
                    consumer.onNext(sliceOf(current), true, streamEnd);
                } else {
                    consumer.onNext(chunk, false, streamEnd);
                }
            });
        }

        void next() {
            if (canSliceCurrent()) {
                consumer.onNext(sliceOf(current), true, streamEnd);
            } else {
                rawStream.next();
            }
        }
    }
```

Reply:
I separated this a bit. I modified `RestBulkAction` to implement the incremental deserialization of doc items.

And then the `IncrementalBulkHandler#Handler` does the actual aggregation of these items. And eventually will do the memory tracking and early firing off `BulkRequests`.